windows/CodePush.cs

```diff
@@ -0,0 +1,668 @@
+﻿using System;
+using System.Collections.Generic;
+using ReactNative.Bridge;
+using ReactNative.Modules.Core;
+using ReactNative.UIManager;
+using Windows.ApplicationModel;
+using Windows.Storage;
+using System.IO;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json;
+using System.Reflection;
+using Windows.Web.Http;
+using Windows.Storage.FileProperties;
+using ReactNative;
+
+namespace CodePush.ReactNative
+{
+    public class CodePushModule : IReactPackage
+    {
+        private static bool needToReportRollback = false;
+        private static bool isRunningBinaryVersion = false;
+        private static bool testConfigurationFlag = false;
+
+        private bool didUpdate = false;
+
+        private string assetsBundleFileName;
+
+        private static readonly string ASSETS_BUNDLE_PREFIX = "ms-appx:///ReactAssets/";
+        private static readonly string BINARY_MODIFIED_TIME_KEY = "binaryModifiedTime";
+        private readonly string CODE_PUSH_PREFERENCES = "CodePush";
+        private static readonly string DEFAULT_JS_BUNDLE_NAME = "index.windows.bundle";
+        private readonly string DOWNLOAD_PROGRESS_EVENT_NAME = "CodePushDownloadProgress";
+        private readonly string FAILED_UPDATES_KEY = "CODE_PUSH_FAILED_UPDATES";
+        private static readonly string FILE_BUNDLE_PREFIX = "ms-appdata:///local";
+        private static readonly string PACKAGE_HASH_KEY = "packageHash";
+        private readonly string PENDING_UPDATE_HASH_KEY = "hash";
+        private readonly string PENDING_UPDATE_IS_LOADING_KEY = "isLoading";
+        private readonly string PENDING_UPDATE_KEY = "CODE_PUSH_PENDING_UPDATE";
+
+        // This needs to be kept in sync with https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/devsupport/DevSupportManager.java#L78
+        private readonly string REACT_DEV_BUNDLE_CACHE_FILE_NAME = "ReactNativeDevBundle.js";
+
+        // Helper classes.
+        private CodePushNativeModule codePushNativeModule;
+        private CodePushPackage codePushPackage;
+
+        // Config properties.
+        private string appVersion;
+        private string deploymentKey;
+        private readonly string serverUrl = "https://codepush.azurewebsites.net/";
+
+        private ReactPage mainPage;
+
+        private static CodePushModule currentInstance;
+
+        public CodePushModule(string deploymentKey, ReactPage mainPage)
+        {
+            codePushPackage = new CodePushPackage();
```

Comment:
I think in the future, we want to enable the scenario that an app has multiple `ReactPage`s, and a user might want to CodePush them individually.